### PR TITLE
Add goodbye messages to Service.Stop

### LINKED ISSUE #4814: Nodes Should Send Goodbye Messages When Shutting Down
Currently when we shut down our node, we just disconnect with all our peers at once. The proper way would be to send a goodbye message to them and then disconnect with them

### PR DESCRIPTION
**Pull request by lsankar4033**
Resolves #4814 

---

# Description

This PR sends goodbye messages to clients when a service is stopped.

### CODE REVIEW

File: beacon-chain/p2p/service.go [hunk 1]
```diff
@@ -277,6 +277,14 @@ func (s *Service) Start() {
 // Stop the p2p service and terminate all peer connections.
 func (s *Service) Stop() error {
 	defer s.cancel()
+
+	for _, pid := range s.Peers().Active() {
+		_, err := s.Send(s.ctx, RPCTopicMappings[RPCGoodByeTopic], RPCGoodByeTopic, pid)
```
Comment (prestonvanloon) [inline beacon-chain/p2p/service.go:282]: This context is already cancelled so i would expect 100% of these to fail.

File: beacon-chain/p2p/service_test.go [hunk 2]
```diff
@@ -100,6 +100,49 @@ func TestService_Stop_DontPanicIfDv5ListenerIsNotInited(t *testing.T) {
...
+		t.Fatal(err)
+	}
+
+	if err := s.Stop(); err != nil {
+		t.Fatal(err)
+	}
+
+	conns := h2.Network().ConnsToPeer(h1.ID())
+	if len(conns) > 0 {
+		t.Error("Peer is still not disconnected despite sending a goodbye message")
```
Comment (prestonvanloon) [inline beacon-chain/p2p/service_test.go:142]: This only tests that a peer was disconnected, not that a goodbye message was sent. Please fix

**Comment by lsankar4033** [comment]
got it.

What's the intended difference b/w sync service and p2p service?

File: beacon-chain/p2p/service.go (hunk 1 above)
Comment (lsankar4033) [reply beacon-chain/p2p/service.go:282]: for my understanding, at what layer of abstraction prior to this is the context cancelled?

Comment (prestonvanloon) [reply beacon-chain/p2p/service.go:282]: I am sorry, I gave incorrect feedback here. The context is not cancelled until this function returns (`defer s.cancel()`)

**Comment by prestonvanloon** [comment]
> What's the intended difference b/w sync service and p2p service?

Sync contains all of the logic with handling pubsub and rpc requests via p2p. 

P2P service contains the broadcast, pubsub, networking configuration implementaitons

**Comment by lsankar4033** [comment]
makes sense! will change

**Comment by lsankar4033** [comment]
hmm, considering that i'm merging to prysmaticlabs:master, i'm not sure why all those other commits showed up. will fix in my fork.

@nisdas does this test look like it's testing the right thing: https://github.com/prysmaticlabs/prysm/pull/5596/commits/0ec680ce1b2b80066def35687428605e76cf72a2 ?

**Comment by nisdas** [comment]
@lsankar4033 that test looks good :+1: , although you might have to rebase your branch it seems like you pulled in some already merged in commits.

**Comment by lsankar4033** [comment]
I'll close this PR. I haven't had time to revisit this and probably won't this week. Probably just best to reassign

**Comment by nisdas** [comment]
No worries @lsankar4033 , thanks for giving this a try :+1:

**Pull request closed without merging**